### ISSUE TITLE
Unable to convert to TypedValue 

### ISSUE THREAD

**Issue by tusharmath** [zio/zio-schema#511]
If the schema is rebuilt from the AST, it doesn't correctly convert a dynamic value back to a typed value. This is specifically happening with records.

```scala
import zio.schema._


final case class User(name: String, age: Int)


implicit val schema = DeriveSchema.gen[User]


val user = User("John", 30)
val userDynamic = schema.toDynamic(user)


println(userDynamic)
println(userDynamic.toTypedValue(schema))
println(userDynamic.toTypedValue(schema.ast.toSchema))

```

https://scastie.scala-lang.org/TJmOjesrS5KY5ZYHOwJTUg

**Comment by pqcfox** [comment]
Going to go ahead and toss my hat in the ring! /attempt #511

**Comment by pqcfox** [comment]
Just dove in to make sure I understand the problem--agree with @andrzejressel on those concerns. Apologies in advance that I'm very new to ZIO, so if anything below doesn't make sense, please let me know!

For the purposes of this MR, @jdegoes would it make sense to add additional typing information to `MetaSchema` and change the `record` method in `Schema` to leverage this (i.e. to construct e.g. a `CaseClass2` via `deriveCaseObject` based on this typing information, rather than a `GenericRecord`)?

The alternatives I thought of were just adding a new, separate `TypedMetaSchema` class that carries more typing information, or the more conservative option of somehow adding warnings when trying to use `toTypedValue` on typeless `Schema`s. 

Any of these seem like good options for me to implement?

**Comment by jdegoes** [comment]
@andrzejressel @pqcfox 

I think the only way to support this would be to do so using reflection. The class name can be stored (indeed, we already have `TypeId`), but then to use it when creating the record seems to require reflection.

An alternative might be to allow passing a 'registry' to `toTypedValue`, something like a `PartialFunction[TypeId, Constructor]`, which would avoid java reflection at the cost of introducing our own version of that.

@tusharmath I know you since moved on but did you have any ideas on how you originally wanted this issue to be solved?

**Comment by tusharmath** [comment]
Hey John! 
I realized that it was not possible to do this a while back. Basically, any custom type can't be decoded this way. We shouldn't allow generating schema of any schema and instead trigger a compile-time error. For eg: `Schema[Schema[List[String]]]` should be allowed but `Schema[Schema[User]]` should not be possible to create ideally. That would be a reasonable developer experience to have, because currently, this introduces runtime bugs.

**Comment by pqcfox** [comment]
Makes sense to me! I can implement that--any list of allowed types I should work off when modifying `Metaschema`?

**Comment by pqcfox** [comment]
(Additionally, if I wanted to address this in a PR, would I do so under this issue, and would it be in scope for the existing bounty?)

**Comment by stanislav-chetvertkov** [comment]
I'm new to zio-schema and just wanted to check if the approach taken in https://github.com/zio/zio-schema/pull/725 makes sense. I naively implemented the registry approach mentioned in https://github.com/zio/zio-schema/issues/511#issuecomment-1880819029. If it does make sense, I'd be happy to gather feedback and finish it.

**Comment by akshatd1007** [comment]
Hi, is this bounty still available to work on?

**Comment by brone1323** [comment]
/attempt #511

Implementation plan: Add a `toTypedValue(schema, registry: Map[TypeId, Schema[_]])` overload to `DynamicValue` that accepts an explicit schema registry. When reconstructing a `DynamicValue.Record` via a `Schema.GenericRecord` (produced by `ast.toSchema`), the registry is consulted by TypeId to retrieve the original typed schema, enabling `construct()` to be called and the round-trip to succeed. The registry approach was suggested by @jdegoes and avoids Java reflection entirely.

**Comment by SAVAMforWIN** [comment]
/attempt #511

Implemented the registry-based direction discussed in the thread, without reflection.

What the patch does:
- adds DynamicValue overloads that accept an explicit Map[TypeId, Schema[_]] registry
- threads that registry through recursive decoding
- when a DynamicValue.Record is decoded against a Schema.GenericRecord coming from schema.ast.toSchema, the decoder now checks the registry for the original nominal Schema.Record
- if found, it reconstructs the typed record instead of stopping at the generic ListMap shape
- if not found, existing behavior is preserved

Validation completed locally with:
java -Xmx2G -jar C:\\Users\\Matteo\\.sbt\\launchers\\1.12.11\\sbt-launch.jar "

**Comment by chengdazhi** [comment]
/attempt 511

Implemented the constructor-registry approach discussed in this issue (https://github.com/zio/zio-schema/issues/511#issuecomment-1880819029): https://github.com/zio/zio-schema/pull/1137

Unlike the naive version, the registry is threaded through the entire conversion, so records are reconstructed at any nesting level (nested records, options, sequences, enum cases, etc.), with fallback to the generic-record behaviour for unregistered types. Tested on Scala 2.12/2.13/3 (JVM + JS), plus MiMa and lint checks.
